### ISSUE TITLE
[java] ImmutableField - false negative when field assigned in constructor conditionally

### ISSUE THREAD

**Issue by MetaBF** [pmd/pmd#3850]
**Affects PMD Version: 6.43.0** 

**Rule: ImmutableField**
 
 https://pmd.github.io/pmd-6.42.0/pmd_rules_java_design.html#immutablefield

**Description:**

Hi, I found a false-negative about rule **ImmutableField**, in this case, the private field `str` is only assigned in the constructor, therefore `str` could be final.

* Code Example 1
```java
public class ExampleImmutableField {

    private String str; // false negative here, could be final

    public ExampleImmutableField(String strLocal, boolean flag) { 
        if (flag){
            this.str = strLocal;
        } else {
            this.str = strLocal+"123";
        }
    }

}
```

I think it's probably due to wrong/limited control-flow analysis since PMD works fine in the following code.

*  Code Example 2
```java
public class ExampleImmutableField {

    private String str; // PMD successfully reports warning here

    public ExampleImmutableField(String strLocal, boolean flag) {
            this.str = strLocal;
    }

} 
```

**Expected outcome:**

PMD should report a violation at line 3 in Code Example 1, but doesn't. This is a false-negative.

**Running PMD through:** *[CLI]*

**Comment by adangel** [comment]
Thanks for the report. I think you are right - in Example 1, the field "str" can be made final and the rule should report this. "str" is a private field whose value never changes once object initialization ends.

Interestingly we have a test case, the manifests the opposite, but I think, the test case is just wrong:
https://github.com/pmd/pmd/blob/59b36b6185e25f25b42c2781a25f81be749ce851/pmd-java/src/test/resources/net/sourceforge/pmd/lang/java/rule/design/xml/ImmutableField.xml#L309-L325

**Comment by ITachiLab** [comment]
@adangel I have a feeling this change is a regression. The rule works well for the ideal condition proposed by @MetaBF, where the field is initialized both inside `if` and `else` blocks. Unfortunately, for the following code the PMD behavior is annoying:

```java
public class ExampleImmutableField {

    private int fooBar;

    public ExampleImmutableField(int newVal, boolean flag) { 
        if (flag) {
            this.fooBar = newVal;
        }
    }
}
```

PMD marks `fooBar` as eligible for `final` modifier, although I really want my `fooBar` field to be 0 by default when the flag is not set because this is the expectation of the algorithm. The above code was valid for PMD 6.12. And sure, I can add `else` block with `this.fooBar = 0` and that will silence PMD for good, but imagine doing this for five, eight, or more fields:

```java
public class ExampleImmutableField {

    private final int a;
    private final int b;
    private final int c;
    private final int d;

    public ExampleImmutableField(int newVal, boolean flag) { 
        if (flag) {
            this.a = magic(newVal);
            this.b = magic(newVal);
            this.c = magic(newVal);
            this.d = magic(newVal);
        } else {
            this.a = 0;
            this.b = 0;
            this.c = 0;
            this.d = 0;
        }
    }
}
```

Or maybe I'm missing something here.
